contrib/babelfishpg_tds/src/backend/tds/tdsbulkload.c

```diff
@@ -642,73 +713,95 @@ SetBulkLoadRowData(TDSRequestBulkLoad request, StringInfo message)
 					offset += 8; /* TODO: Ignored the Data Text TimeStamp for now. */
 					request->currentBatchSize += 8;
 
-					memcpy(&rowData->columnValues[i].len, &message->data[offset], sizeof(uint32_t));
+					memcpy(&len, &message->data[offset], sizeof(uint32_t));
 					offset +=  sizeof(uint32_t);
 					request->currentBatchSize += sizeof(uint32_t);
-					if (rowData->columnValues[i].len == 0) /* null */
+					if (len == 0) /* null */
 					{
 						rowData->isNull[i] = true;
 						i++;
 						continue;
 					}
 
-					CheckForInvalidLength(rowData, request, i);
+					CheckForInvalidLength(len, request, i);
 
-					if (rowData->columnValues[i].len > rowData->columnValues[i].maxlen)
-						enlargeStringInfo(&rowData->columnValues[i], rowData->columnValues[i].len);
+					CheckMessageHasEnoughBytesToRead(&message, len);
 
-					CheckMessageHasEnoughBytesToRead(&message, rowData->columnValues[i].len);
+					/* Build temp Stringinfo. */
+					temp->data = &message->data[offset];
+					temp->len = len;
+					temp->maxlen = colmetadata[i].maxLen;
+					temp->cursor = 0;
 
-					memcpy(rowData->columnValues[i].data, &message->data[offset], rowData->columnValues[i].len);
-					offset += rowData->columnValues[i].len;
-					request->currentBatchSize += rowData->columnValues[i].len;
+					/* Create and store the appropriate datum for this column. */
+					switch(colmetadata[i].columnTdsType)
+					{
+						case TDS_TYPE_TEXT:
+							rowData->columnValues[i] = TdsTypeVarcharToDatum(temp, colmetadata[i].collation, colmetadata[i].columnTdsType);
+						break;
+						case TDS_TYPE_NTEXT:
+							rowData->columnValues[i] = TdsTypeNCharToDatum(temp);
+						break;
+						case TDS_TYPE_IMAGE:
+							rowData->columnValues[i] = TdsTypeVarbinaryToDatum(temp);
+						break;
+					}
+
+					offset += len;
+					request->currentBatchSize += len;
 				}
 				break;
 				case TDS_TYPE_XML:
 				{
-					StringInfo plpStr;
-					ParameterToken temp = palloc0(sizeof(ParameterTokenData));
-					retStatus = ReadBcpPlp(temp, &message, request);
+					ParameterToken token = palloc0(sizeof(ParameterTokenData));
+
+					retStatus = ReadBcpPlp(token, &message, request);
 					CheckPLPStatusNotOK(request, retStatus, i);
-					if (temp->isNull) /* null */
+					if (token->isNull) /* null */
 					{
 						rowData->isNull[i] = true;
 						i++;
-						temp->isNull = false;
+						token->isNull = false;
 						continue;
 					}
 
-					plpStr = TdsGetPlpStringInfoBufferFromToken(message->data, temp);
-					rowData->columnValues[i] = *plpStr;
-					pfree(plpStr);
+					temp = TdsGetPlpStringInfoBufferFromToken(message->data, token);
```

Comment:
Here again we're overwriting the pointer. It's better to define temp variable in each switch case where you need the same.

Reply:
If we allocate stringInfo inside the switch case which happens in a loop for all rows' columns', it would defeat the purpose of the optimisation done in this PR. Rather just freed temp just before it